### PR TITLE
Add support for "phantom" BOLT 12 offers, up to the invoice_request step

### PR DESCRIPTION
**Pull request by TheBlueMatt**
In the BOLT 11 world, we have specific support for what we call
"phantom nodes" - creating invoices which can be paid to any one of
a number of nodes by adding route-hints which represent nodes that
do not exist.

In BOLT 12, blinded paths make a similar feature much simpler - we
can simply add blinded paths which terminate at different nodes.
The blinding means that the sender is none the wiser.

Here we add logic to fetch an `OfferBuilder` which can generate an
offer payable to any one of a set of nodes. We retain the "phantom"
terminology even though there are no longer any "phantom" nodes.

Note that the current logic only supports the `invoice_request`
message going to any of the participating nodes, it then replies
with a `Bolt12Invoice` which can only be paid to the responding
node. Future work may relax this restriction.

Also,

> Note that we will not yet use the blinded payment path phantom
support which requires additional future work. However, allowing
them to be authenticated in a phantom configuration should allow
for compatibility across versions once the building logic lands.

Progress towards full solution for #4313, but enough to get folks started.

### CODE REVIEW

File: lightning/src/crypto/streams.rs [hunk 1]
```diff
@@ -87,31 +87,34 @@ pub(crate) fn chachapoly_encrypt_with_swapped_aad(
...
+/// accepting any being correct as sufficient.
 ///
-/// Note that we do *not* use the provided AAD as the standard ChaCha20Poly1305 AAD as that would
+/// Note that we do *not* use the provided AADs as the standard ChaCha20Poly1305 AAD as that would
 /// require placing it first and prevent us from avoiding redundant Poly1305 rounds. Instead, the
 /// ChaCha20Poly1305 MAC check is tweaked to move the AAD to *after* the the contents being
 /// checked, effectively treating the contents as the AAD for the AAD-containing MAC but behaving
 /// like classic ChaCha20Poly1305 for the non-AAD-containing MAC.
-pub(crate) struct ChaChaDualPolyReadAdapter<R: Readable> {
+pub(crate) struct ChaChaTriPolyReadAdapter<R: Readable> {
```
Comment (joostjager) [inline lightning/src/crypto/streams.rs:111]: Would you say there is sufficient test coverage on this? I think there is some higher-level coverage, but no direct unit test.

File: lightning/src/crypto/streams.rs [hunk 2]
```diff
@@ -87,31 +87,34 @@ pub(crate) fn chachapoly_encrypt_with_swapped_aad(
...
 	pub readable: R,
-	pub used_aad: bool,
+	pub used_aad_a: bool,
+	pub used_aad_b: bool,
```
Comment (joostjager) [inline lightning/src/crypto/streams.rs]: Both being true is never going to happen. Can you use a data type that has three possible values?

Can also consider using local and phantom here in the names, as this adapter isn't used for anything else.

File: lightning/src/blinded_path/payment.rs [hunk 3]
```diff
@@ -284,15 +284,17 @@ impl BlindedPaymentPath {
...
+		let read_arg = (rho, receive_auth_key.0, phantom_auth_key);
+
 		let encrypted_control_tlvs =
 			&self.inner_path.blinded_hops.get(0).ok_or(())?.encrypted_payload;
 		let mut s = Cursor::new(encrypted_control_tlvs);
 		let mut reader = FixedLengthReader::new(&mut s, encrypted_control_tlvs.len() as u64);
-		let ChaChaDualPolyReadAdapter { readable, used_aad } =
-			ChaChaDualPolyReadAdapter::read(&mut reader, (rho, receive_auth_key.0))
-				.map_err(|_| ())?;
+		let ChaChaTriPolyReadAdapter { readable, used_aad_a, used_aad_b } =
```
Comment (joostjager) [inline lightning/src/blinded_path/payment.rs]: Names like `used_local` and `used_phantom` could be helpful

File: lightning/src/ln/msgs.rs [hunk 4]
```diff
@@ -3723,13 +3726,14 @@ where
 							features,
 							next_blinding_override,
 						}),
-					used_aad,
+					used_aad_a,
+					used_aad_b,
```
Comment (joostjager) [inline lightning/src/ln/msgs.rs]: local/phantom in name?

File: lightning/src/ln/channelmanager.rs [hunk 5]
```diff
@@ -14045,6 +14084,38 @@ where
...
+			if peer_chans.iter().any(|chan| chan.is_usable)
+				&& peer_chans.iter().any(|c| c.counterparty.features.supports_onion_messages())
+			{
+				res.push(MessageForwardNode {
+					node_id: peer_chans[0].counterparty.node_id,
+					short_channel_id: peer_chans
+						.iter()
+						.filter(|chan| chan.is_usable)
+						.filter_map(|chan| chan.short_channel_id)
+						.min(),
```
Comment (joostjager) [inline lightning/src/ln/channelmanager.rs]: Why min?

File: lightning/src/ln/channelmanager.rs [hunk 6]
```diff
@@ -14045,6 +14084,38 @@ where
...
+	) -> Vec<MessageForwardNode> {
+		channel_list.sort_unstable_by_key(|chan| chan.counterparty.node_id);
+		let mut res = Vec::new();
+		// TODO: When MSRV reaches 1.77 use chunk_by
+		let mut start = 0;
+		while start < channel_list.len() {
+			let counterparty_node_id = channel_list[start].counterparty.node_id;
+			let end = channel_list[start..]
+				.iter()
+				.position(|chan| chan.counterparty.node_id != counterparty_node_id)
```
Comment (joostjager) [inline lightning/src/ln/channelmanager.rs:14075]: is this correct? end is a position relative to start, but below, it is used as an absolute index?

File: lightning/src/ln/channelmanager.rs [hunk 7]
```diff
@@ -14045,6 +14084,38 @@ where
 		now
 	}
 
+	fn channel_details_to_forward_node(
```
Comment (joostjager) [inline lightning/src/ln/channelmanager.rs]: Some doc on the fn would be nice

File: lightning/src/ln/functional_test_utils.rs [hunk 8]
```diff
@@ -4404,21 +4404,41 @@ pub fn create_chanmon_cfgs(node_count: usize) -> Vec<TestChanMonCfg> {
...
 		let tx_broadcaster = test_utils::TestBroadcaster::new(Network::Testnet);
 		let fee_estimator = test_utils::TestFeeEstimator::new(253);
 		let chain_source = test_utils::TestChainSource::new(Network::Testnet);
 		let logger = test_utils::TestLogger::with_id(format!("node {}", i));
 		let persister = test_utils::TestPersister::new();
-		let seed = [i as u8; 32];
-		let keys_manager = if predefined_keys_ids.is_some() {
+		let mut seed = [i as u8; 32];
+		if phantom {
+			// We would ideally randomize keys on every test run, but some tests fail in that case.
```
Comment (joostjager) [inline lightning/src/ln/functional_test_utils.rs:4429]: Why is that ideal? Aren't non-deterministic tests ideal?

File: lightning/src/blinded_path/payment.rs [hunk 9]
```diff
@@ -284,15 +284,17 @@ impl BlindedPaymentPath {
 			node_signer.ecdh(Recipient::Node, &self.inner_path.blinding_point, None)?;
 		let rho = onion_utils::gen_rho_from_shared_secret(&control_tlvs_ss.secret_bytes());
 		let receive_auth_key = node_signer.get_receive_auth_key();
+		let phantom_auth_key = node_signer.get_expanded_key().phantom_node_blinded_path_key;
```
Comment (joostjager) [inline lightning/src/blinded_path/payment.rs:271]: You could have None here in case the node isn't configured for phantom payments. I think that might help keep that case in mind for readers, and also saves some computation?

File: lightning/src/ln/offers_tests.rs [hunk 10]
```diff
@@ -2571,3 +2577,111 @@ fn no_double_pay_with_stale_channelmanager() {
 	// generated in response to the duplicate invoice.
 	assert!(nodes[0].node.get_and_clear_pending_events().is_empty());
 }
+
+#[test]
+fn creates_and_pays_for_phantom_offer() {
+	// XXX: share expanded key
```
Comment (joostjager) [inline lightning/src/ln/offers_tests.rs]: XXX to be resolved

File: lightning/src/ln/offers_tests.rs [hunk 11]
```diff
@@ -2571,3 +2577,111 @@ fn no_double_pay_with_stale_channelmanager() {
...
+	nodes[0].onion_messenger.handle_onion_message(node_c_id, &onion_message);
+
+	let (invoice, _) = extract_invoice(&nodes[0], &onion_message);
+	assert_eq!(invoice.amount_msats(), 10_000_000);
+
+	route_bolt12_payment(&nodes[0], &[&nodes[2]], &invoice);
+	expect_recent_payment!(&nodes[0], RecentPaymentDetails::Pending, payment_id);
+
+	claim_bolt12_payment(&nodes[0], &[&nodes[2]], payment_context, &invoice);
+	expect_recent_payment!(&nodes[0], RecentPaymentDetails::Fulfilled, payment_id);
```
Comment (joostjager) [inline lightning/src/ln/offers_tests.rs]: Can the two payments happen in a loop to avoid duplication?

File: lightning/src/ln/offers_tests.rs [hunk 12]
```diff
@@ -2571,3 +2577,111 @@ fn no_double_pay_with_stale_channelmanager() {
...
+	let (invoice, _) = extract_invoice(&nodes[0], &onion_message);
+	assert_eq!(invoice.amount_msats(), 10_000_000);
+
+	route_bolt12_payment(&nodes[0], &[&nodes[1]], &invoice);
+	expect_recent_payment!(&nodes[0], RecentPaymentDetails::Pending, payment_id);
+
+	claim_bolt12_payment(&nodes[0], &[&nodes[1]], payment_context, &invoice);
+	expect_recent_payment!(&nodes[0], RecentPaymentDetails::Fulfilled, payment_id);
+
+	// Then pay again via node C
```
Comment (joostjager) [inline lightning/src/ln/offers_tests.rs]: Note that this is the critical part. Node C knows nothing about the offer, but is going to respond to the invoice request. Emphasize the cool part :)

File: lightning/src/ln/channelmanager.rs [hunk 13]
```diff
@@ -13416,6 +13416,45 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
...
+	/// `other_nodes_channels.len() + 1`) not all nodes will participate in receiving funds.
+	/// Because the parameterized [`MessageRouter`] will only get a chance to limit the number of
+	/// paths *per-node*, it is important to set this for offers which will be included in a QR
+	/// code.
+	///
+	/// See [`Self::create_offer_builder`] for more details on the blinded path construction.
+	///
+	/// [`ExpandedKey`]: inbound_payment::ExpandedKey
+	pub fn create_phantom_offer_builder(
+		&$self, other_nodes_channels: Vec<(PublicKey, Vec<ChannelDetails>)>,
```
Comment (joostjager) [inline lightning/src/ln/channelmanager.rs:13429]: Wouldn't passing in `MessageForwardNode` directly create more flexibility?

File: lightning/src/crypto/streams.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/crypto/streams.rs:111]: I think so? I mean if its horribly broken the higher-level code will also break. Its also somewhat indirectly fuzzed through the onion decoding fuzzing.

File: lightning/src/blinded_path/payment.rs (hunk 3 above)
Comment (TheBlueMatt) [reply lightning/src/blinded_path/payment.rs]: Sadly I can't rename the fields since I switched to an enum.

File: lightning/src/ln/msgs.rs (hunk 4 above)
Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs]: Sadly I can't rename the fields since I switched to an enum.

File: lightning/src/ln/channelmanager.rs (hunk 5 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: Its the oldest, matching the existing logic we have for the non-phantom case.

File: lightning/src/ln/channelmanager.rs (hunk 7 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: It converts `ChannelDetails` to `ForwardNode`s. Do you have some specific context or call semantics you feel like is missing from the name?

File: lightning/src/ln/functional_test_utils.rs (hunk 8 above)
Comment (TheBlueMatt) [reply lightning/src/ln/functional_test_utils.rs:4429]: We've historically had tests that accidentally pass because of key ordering or hashmap ordering, which can hide bugs. IMO tests should be robust against basic things like that (that might change in future PRs anyway!)

File: lightning/src/blinded_path/payment.rs (hunk 9 above)
Comment (TheBlueMatt) [reply lightning/src/blinded_path/payment.rs:271]: Yea, we don't currently have a separate method on the signer that returns `None` in cases where we aren't doing phantom so we'd have to add a new key-fetcher (or `bool`-fetcher) to decide whether to do phantom validation. I thought about doing that but it seemed way cleaner to reuse the `ExpandedKey` (which is already used for all the inbound-payment key logic, including phantom). Its also not very much compute at all to just do one round of poly1305 + the finish logic so I'm not sure its worth the extra work.

File: lightning/src/ln/channelmanager.rs (hunk 13 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:13429]: Hmm, in theory, but I'm not quite sure what the use-case for it would be? Someone could already filter peers by filtering the `ChannelDetails`. I guess it would let someone include a `MessageForwardNode` for a peer they don't have a channel with, but that seems a bit strange to do especially for a phantom setup. Avoiding it means we don't have to add a new public method on `ChannelManager` which seems marginally nicer than not.

File: lightning/src/crypto/streams.rs (hunk 1 above)
Comment (joostjager) [reply lightning/src/crypto/streams.rs:111]: I think lower level unit tests do have value even if there are higher level tests already.

File: lightning/src/ln/channelmanager.rs (hunk 13 above)
Comment (joostjager) [reply lightning/src/ln/channelmanager.rs:13429]: This seems fine indeed.  Wanted to bring up the alternative just in case.

File: lightning/src/blinded_path/payment.rs (hunk 3 above)
Comment (joostjager) [reply lightning/src/blinded_path/payment.rs]: Why not use `Local` and `Phantom` in the enum? This adapter isn't used for anything else.

File: lightning/src/ln/channelmanager.rs (hunk 5 above)
Comment (joostjager) [reply lightning/src/ln/channelmanager.rs]: Is that to maximize the chances that it still exist? Curious where the existing logic is located also.

File: lightning/src/ln/channelmanager.rs (hunk 7 above)
Comment (joostjager) [reply lightning/src/ln/channelmanager.rs]: Something like extracting the unique nodes with their oldest channel.

File: lightning/src/ln/functional_test_utils.rs (hunk 8 above)
Comment (joostjager) [reply lightning/src/ln/functional_test_utils.rs:4429]: I think there are pros and cons. Fully deterministic tests are desirable too sometimes.

File: lightning/src/ln/offers_tests.rs (hunk 11 above)
Comment (joostjager) [reply lightning/src/ln/offers_tests.rs]: Nice cleanup

File: lightning/src/ln/offers_tests.rs (hunk 12 above)
Comment (joostjager) [reply lightning/src/ln/offers_tests.rs]: Nice addition to the comment.

File: lightning/src/crypto/streams.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/crypto/streams.rs:111]: Not really sure what you'd want to test? That it can decrypt correctly? That's definitely well-covered. The one bug we did have in it was found by our fuzz test. More test coverage is great but more tests to get the same coverage seems like a net-negative.

File: lightning/src/blinded_path/payment.rs (hunk 3 above)
Comment (TheBlueMatt) [reply lightning/src/blinded_path/payment.rs]: Because that's really confusing when you read the stream type that isn't dedicated to only use in blinded paths. I'm not at all convinced the callsites are unclear, and we also might need to/want to upstream this at some point to `rust-bitcoin`.

File: lightning/src/ln/channelmanager.rs (hunk 5 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: Yea, basically. Absent any better criteria we assume that the oldest channel is most likely to stick around. We do it in `get_peers_for_blinded_path` as well using `funded_channel.context.channel_creation_height`

File: lightning/src/ln/functional_test_utils.rs (hunk 8 above)
Comment (TheBlueMatt) [reply lightning/src/ln/functional_test_utils.rs:4429]: Indeed there's tradeoffs. In this case its not a behavior change so there aren't many :p

File: lightning/src/ln/inbound_payment.rs [hunk 14]
```diff
@@ -58,6 +58,10 @@ pub struct ExpandedKey {
 	/// The key used to authenticate spontaneous payments' metadata as previously registered with LDK
 	/// for inclusion in a blinded path.
 	spontaneous_pmt_key: [u8; 32],
+	/// The key used to authenticate phantom-node-shared blinded paths as generated by us. Note
+	/// that this is not used for blinded paths which are not expected to be shared across nodes
```
Comment (jkczyz) [inline lightning/src/ln/inbound_payment.rs]: s/which/that

File: lightning/src/crypto/streams.rs [hunk 15]
```diff
@@ -87,31 +87,34 @@ pub(crate) fn chachapoly_encrypt_with_swapped_aad(
 /// Enables the use of the serialization macros for objects that need to be simultaneously decrypted
 /// and deserialized. This allows us to avoid an intermediate Vec allocation.
 ///
-/// This variant of [`ChaChaPolyReadAdapter`] calculates Poly1305 tags twice, once using the given
-/// key and once with the given 32-byte AAD appended after the encrypted stream, accepting either
-/// being correct as sufficient.
+/// This variant of [`ChaChaPolyReadAdapter`] calculates Poly1305 tags thrice, once using the given
+/// key and twice with each of the two given 32-byte AADs appended after the encrypted stream,
```
Comment (jkczyz) [inline lightning/src/crypto/streams.rs]: s/twice with each of the two/once each for the two

File: lightning/src/crypto/streams.rs [hunk 16]
```diff
@@ -84,6 +84,18 @@ pub(crate) fn chachapoly_encrypt_with_swapped_aad(
...
+#[derive(PartialEq, Eq)]
+pub(crate) enum TriPolyAADUsed {
+	/// No AAD was used.
+	///
+	/// The HMAC validated with standard ChaCha20Poly1305.
+	NoAAD,
+	/// The HMAC vlidated using the first AAD provided.
+	A,
+	/// The HMAC vlidated using the second AAD provided.
+	B,
```
Comment (jkczyz) [inline lightning/src/crypto/streams.rs]: `FirstAAD` and `SecondAAD` to match `NoAAD`?

File: lightning/src/ln/channelmanager.rs [hunk 17]
```diff
@@ -13416,6 +13416,45 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
...
+	/// [`ChannelManager`] (or [`OffersMessageFlow`]) using the same [`ExpandedKey`] (as returned
+	/// from [`NodeSigner::get_expanded_key`]). This allows any nodes participating in a BOLT 11
+	/// "phantom node" cluster to also receive BOLT 12 payments.
+	///
+	/// Note that, unlike with BOLT 11 invoices, BOLT 12 "phantom" offers do not in fact have any
+	/// "phantom node" appended to receiving paths. Instead, multiple blinded paths are simply
+	/// included which terminate at different final nodes.
+	///
+	/// `other_nodes_channels` must be set to a list of each participating node's `node_id` (from
+	/// [`NodeSigner::get_node_id`] with a [`Recipient::Node`]) and its channels.
```
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs:13416]: Could you update `NodeSigner::get_node_id` to mention that `Recipient::Node` should be used for BOLT12 "phantom" payments?

File: lightning/src/ln/channelmanager.rs [hunk 18]
```diff
@@ -13416,6 +13416,45 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
...
+	///
+	/// `path_count_limit` is used to limit the number of blinded paths included in the resulting
+	/// [`Offer`]. Note that if this is less than the number of participating nodes (i.e.
+	/// `other_nodes_channels.len() + 1`) not all nodes will participate in receiving funds.
+	/// Because the parameterized [`MessageRouter`] will only get a chance to limit the number of
+	/// paths *per-node*, it is important to set this for offers which will be included in a QR
```
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs]: s/which/that

**Comment by TheBlueMatt** [comment]
Rebased without further changes.

File: lightning/src/ln/channelmanager.rs (hunk 5 above)
Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs]: Won't an old channel that is spliced recently seem newer? Also, should we use `inbound_scid_alias` if set as is done in `get_peers_for_blinded_path` by using `FundedChannel::get_inbound_scid`?

File: lightning/src/ln/channelmanager.rs (hunk 13 above)
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs:13429]: For BOLT11 phantom, we have the invoice creator pass all the details, even their own. Should we do the same here? More broadly, are we expecting the caller to query all the nodes for the details? Or do we expect them to have some process that periodically does that and they are just looking up the results?

File: lightning/src/crypto/streams.rs (hunk 16 above)
Comment (TheBlueMatt) [reply lightning/src/crypto/streams.rs]: Dropped the `AAD` suffix but did this.

File: lightning/src/ln/channelmanager.rs (hunk 17 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:13416]: Not quite sure what you're referring to? We don't use `get_node_id` directly, except via blinded path creation in cases where we aren't using blinded paths.

File: lightning/src/ln/channelmanager.rs (hunk 5 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: > Won't an old channel that is spliced recently seem newer?

Indeed it will. There's no obvious fix for that AFAICT, but it also shouldn't really matter - we could just pick a random SCID too if we want...

> Also, should we use inbound_scid_alias if set as is done in get_peers_for_blinded_path by using FundedChannel::get_inbound_scid?

Oops, right.

File: lightning/src/ln/channelmanager.rs (hunk 13 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:13429]: > For BOLT11 phantom, we have the invoice creator pass all the details, even their own. Should we do the same here?

I don't really have a strong opinion. For BOLT 11 we use a loose method not on `ChannelManager` but here I wanted to use the offers flow for simplicity. We could make it a loose method on the flow object and make the dev pass in all nodes' details or we could do so on `ChannelManager`, it just felt pretty weird to require someone pass in local details directly to the `ChannelManager` they just queried.

 > More broadly, are we expecting the caller to query all the nodes for the details? Or do we expect them to have some process that periodically does that and they are just looking up the results?

Probably the second? as long as its relatively often-refreshed.

Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:13429]: If the second, then it seems the caller would need to fetch them and then filter their own, though? Unless the request says to filter by some node id.

File: lightning/src/ln/channelmanager.rs (hunk 5 above)
Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs]: Could we expose `channel_creation_height` in `ChannelDetails`?

File: lightning/src/ln/channelmanager.rs (hunk 13 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:13429]: indeed, they would need to filter their own. We could of course automate that step. Do you have a strong opinion on the design here?

Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:13429]: Not a strong opinion, though naming-wise using `peers` here and `per_node_peers` in `flow.rs` isn't accurate and at least should be clarified in the latter case.

Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:13429]: I guess technically they are a set of peers, so feel free to disregard.

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:13429]: I went ahead and at least skipped adding our own peers duplicatively if the caller already provided them.

**Comment by TheBlueMatt** [comment]
Squashed and fixed remaining rustfmt and CI issues:

```diff
$ git diff-tree -U1 5a672bdad 10391b710
diff --git a/ext-functional-test-demo/src/main.rs b/ext-functional-test-demo/src/main.rs
index 654cf91e01..67eb8c776f 100644
--- a/ext-functional-test-demo/src/main.rs
+++ b/ext-functional-test-demo/src/main.rs
@@ -19,2 +19,3 @@ mod tests {
 			&self, _seed: &[u8; 32], _now: Duration, _v2_remote_key_derivation: bool,
+			_phantom_seed: Option<&[u8; 32]>,
 		) -> Box<dyn DynKeysInterfaceTrait<EcdsaSigner = DynSigner>> {
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 0da4c52dc0..64cbc92a22 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -13432,3 +13432,5 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
 		let mut peers = Vec::with_capacity(other_nodes_channels.len() + 1);
-		peers.push(($self.get_our_node_id(), $self.get_peers_for_blinded_path()));
+		if !other_nodes_channels.iter().any(|(node_id, _)| *node_id == $self.get_our_node_id()) {
+			peers.push(($self.get_our_node_id(), $self.get_peers_for_blinded_path()));
+		}
 		for (node_id, peer_chans) in other_nodes_channels {
diff --git a/lightning/src/onion_message/packet.rs b/lightning/src/onion_message/packet.rs
index 414232f059..cd9a923b07 100644
--- a/lightning/src/onion_message/packet.rs
+++ b/lightning/src/onion_message/packet.rs
@@ -337,6 +337,3 @@ impl<H: CustomOnionMessageHandler + ?Sized, L: Logger + ?Sized>
 			None => return Err(DecodeError::InvalidValue),
-			Some(ChaChaTriPolyReadAdapter {
-				readable: ControlTlvs::Forward(tlvs),
-				used_aad,
-			}) => {
+			Some(ChaChaTriPolyReadAdapter { readable: ControlTlvs::Forward(tlvs), used_aad }) => {
 				if used_aad != TriPolyAADUsed::None || message_type.is_some() {
@@ -346,18 +343,14 @@ impl<H: CustomOnionMessageHandler + ?Sized, L: Logger + ?Sized>
 			},
-			Some(ChaChaTriPolyReadAdapter {
-				readable: ControlTlvs::Dummy,
-				used_aad,
-			}) => Ok(Payload::Dummy {
-				control_tlvs_authenticated: used_aad != TriPolyAADUsed::None,
-			}),
-			Some(ChaChaTriPolyReadAdapter {
-				readable: ControlTlvs::Receive(tlvs),
-				used_aad,
-			}) => Ok(Payload::Receive {
-				control_tlvs: ReceiveControlTlvs::Unblinded(tlvs),
-				reply_path,
-				message: message.ok_or(DecodeError::InvalidValue)?,
-				control_tlvs_from_local_node: used_aad == TriPolyAADUsed::First,
-				control_tlvs_from_phantom_participant: used_aad == TriPolyAADUsed::Second,
-			}),
+			Some(ChaChaTriPolyReadAdapter { readable: ControlTlvs::Dummy, used_aad }) => {
+				Ok(Payload::Dummy { control_tlvs_authenticated: used_aad != TriPolyAADUsed::None })
+			},
+			Some(ChaChaTriPolyReadAdapter { readable: ControlTlvs::Receive(tlvs), used_aad }) => {
+				Ok(Payload::Receive {
+					control_tlvs: ReceiveControlTlvs::Unblinded(tlvs),
+					reply_path,
+					message: message.ok_or(DecodeError::InvalidValue)?,
+					control_tlvs_from_local_node: used_aad == TriPolyAADUsed::First,
+					control_tlvs_from_phantom_participant: used_aad == TriPolyAADUsed::Second,
+				})
+			},
 		}
```

File: lightning/src/ln/channelmanager.rs (hunk 5 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: I really don't think it matters, we're basically flipping a coin here, but I went ahead and changed both to "highest local balance" cause it seems marginally better than "oldest" (or "newest" or "most recently spliced").

**Comment by TheBlueMatt** [comment]
Apologies, I'd forgotten about that comment, I pushed an update to the criteria that I think makes marginally more sense.
